### PR TITLE
[LLM Runtime] refactor itrex backend based on the latest Jblas

### PR DESCRIPTION
**Pull request by zhewang1-intc**
## Type of Change

feature or bug fix or documentation or others
API changed or not

## Description

detail description 
JIRA ticket: xxx

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

File: intel_extension_for_transformers/llm/library/.clang-format
Comment (DDEle) [inline intel_extension_for_transformers/llm/library/.clang-format:1]: Delete it as:
- Developer should sync from the jblas internal repo instead if directly editing jblas in ITREX
- It may affect other libraries

**Review by DDEle — COMMENTED** [review]
Glad to see the great refactor!

(seem someone need to fix the cpp graph)

File: intel_extension_for_transformers/llm/library/.clang-format
Comment (zhewang1-intc) [reply intel_extension_for_transformers/llm/library/.clang-format:1]: fix in [376c6c3](https://github.com/intel/intel-extension-for-transformers/pull/769/commits/376c6c370d1d355c8535e51bea7e579373a506ce)
BTW sometimes we need to temporarily modify some Jblas code in itrex to verify our ideas. Without the clang-format, git diff would be a nightmare :sweat_smile:

**Comment by luoyu-intel** [comment]
CompType=fp32  1.3x speedup
main branch:
perf_total_per_op_us[                     ADD] =   0.153 ms
perf_total_per_op_us[                     MUL] =   0.231 ms
perf_total_per_op_us[                RMS_NORM] =   0.399 ms
perf_total_per_op_us[                 RESHAPE] =   0.092 ms
perf_total_per_op_us[                    VIEW] =   0.176 ms
perf_total_per_op_us[                 PERMUTE] =   0.029 ms
perf_total_per_op_us[               TRANSPOSE] =   0.023 ms
perf_total_per_op_us[                GET_ROWS] =   0.013 ms
perf_total_per_op_us[                    ROPE] =   0.955 ms
perf_total_per_op_us[                FFN_SILU] =  16.995 ms
perf_total_per_op_us[              FLASH_ATTN] =   0.665 ms
perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   0.842 ms
perf_total_per_op_us[           INNER PRODUCT] =  13.321 ms

pr branch:
perf_total_per_op_us[                     ADD] =   0.147 ms
perf_total_per_op_us[                     MUL] =   0.247 ms
perf_total_per_op_us[                RMS_NORM] =   0.418 ms
perf_total_per_op_us[                 RESHAPE] =   0.104 ms
perf_total_per_op_us[                    VIEW] =   0.273 ms
perf_total_per_op_us[                 PERMUTE] =   0.021 ms
perf_total_per_op_us[               TRANSPOSE] =   0.018 ms
perf_total_per_op_us[                GET_ROWS] =   0.007 ms
perf_total_per_op_us[                    ROPE] =   1.605 ms
perf_total_per_op_us[                 MUL_QKV] =   5.460 ms
perf_total_per_op_us[                FFN_SILU] =  12.781 ms
perf_total_per_op_us[              FLASH_ATTN] =   0.947 ms
perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   1.567 ms
perf_total_per_op_us[           INNER PRODUCT] =   2.931 ms

**Comment by luoyu-intel** [comment]
CompType=int8  1.08x
main:
perf_total_per_op_us[                     ADD] =   0.138 ms
perf_total_per_op_us[                     MUL] =   0.179 ms
perf_total_per_op_us[                RMS_NORM] =   0.386 ms
perf_total_per_op_us[                 RESHAPE] =   0.077 ms
perf_total_per_op_us[                    VIEW] =   0.241 ms
perf_total_per_op_us[                 PERMUTE] =   0.020 ms
perf_total_per_op_us[               TRANSPOSE] =   0.025 ms
perf_total_per_op_us[                GET_ROWS] =   0.013 ms
perf_total_per_op_us[                    ROPE] =   1.045 ms
perf_total_per_op_us[                 MUL_QKV] =   4.812 ms
perf_total_per_op_us[                FFN_SILU] =  11.572 ms
perf_total_per_op_us[              FLASH_ATTN] =   0.629 ms
perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   1.036 ms
perf_total_per_op_us[           INNER PRODUCT] =   2.455 ms


pr branch:
perf_total_per_op_us[                     ADD] =   0.129 ms
perf_total_per_op_us[                     MUL] =   0.174 ms
perf_total_per_op_us[                RMS_NORM] =   0.366 ms
perf_total_per_op_us[                 RESHAPE] =   0.094 ms
perf_total_per_op_us[                    VIEW] =   0.252 ms
perf_total_per_op_us[                 PERMUTE] =   0.020 ms
perf_total_per_op_us[               TRANSPOSE] =   0.021 ms
perf_total_per_op_us[                GET_ROWS] =   0.011 ms
perf_total_per_op_us[                    ROPE] =   1.017 ms
perf_total_per_op_us[                 MUL_QKV] =   4.343 ms
perf_total_per_op_us[                FFN_SILU] =  10.654 ms
perf_total_per_op_us[              FLASH_ATTN] =   0.617 ms
perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   1.018 ms
perf_total_per_op_us[           INNER PRODUCT] =   2.180 ms

**Comment by luoyu-intel** [comment]
comp_type=bf16, ~1.3x 
main(no ffn, qkv fusion) :
perf_total_per_op_us[                     ADD] =   0.187 ms
perf_total_per_op_us[                     MUL] =   0.530 ms
perf_total_per_op_us[                    SILU] =   1.878 ms
perf_total_per_op_us[                RMS_NORM] =   0.487 ms
perf_total_per_op_us[                 RESHAPE] =   0.116 ms
perf_total_per_op_us[                    VIEW] =   0.218 ms
perf_total_per_op_us[                 PERMUTE] =   0.028 ms
perf_total_per_op_us[               TRANSPOSE] =   0.029 ms
perf_total_per_op_us[                GET_ROWS] =   0.008 ms
perf_total_per_op_us[                    ROPE] =   0.785 ms
perf_total_per_op_us[              FLASH_ATTN] =   0.487 ms
perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   0.714 ms
perf_total_per_op_us[           INNER PRODUCT] =  41.611 ms

pr branch:
perf_total_per_op_us[                     ADD] =   0.164 ms
perf_total_per_op_us[                     MUL] =   0.286 ms
perf_total_per_op_us[                RMS_NORM] =   0.433 ms
perf_total_per_op_us[                 RESHAPE] =   0.095 ms
perf_total_per_op_us[                    VIEW] =   0.276 ms
perf_total_per_op_us[                 PERMUTE] =   0.021 ms
perf_total_per_op_us[               TRANSPOSE] =   0.023 ms
perf_total_per_op_us[                GET_ROWS] =   0.007 ms
perf_total_per_op_us[                    ROPE] =   0.786 ms
perf_total_per_op_us[                 MUL_QKV] =   6.667 ms
perf_total_per_op_us[                FFN_SILU] =  16.416 ms
perf_total_per_op_us[              FLASH_ATTN] =   0.550 ms
perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   0.692 ms
perf_total_per_op_us[           INNER PRODUCT] =   3.089 ms

**Comment by luoyu-intel** [comment]
long prompt CompType=int8, 1.2x:

main:
model_print_timings: prompt eval time =  1430.17 ms /  2023 tokens (    0.71 ms per token)
model_print_timings:        eval time =   410.02 ms /    15 runs   (   27.33 ms per token)
model_print_timings:       total time =  2969.41 ms
========== eval time log of each prediction ==========
prediction   0, time: 1430.17ms
prediction   1, time: 28.56ms
prediction   2, time: 27.65ms
prediction   3, time: 27.34ms
prediction   4, time: 27.46ms
prediction   5, time: 27.33ms
prediction   6, time: 27.24ms
prediction   7, time: 27.28ms
prediction   8, time: 27.18ms
prediction   9, time: 27.25ms
prediction  10, time: 27.35ms
prediction  11, time: 27.12ms
prediction  12, time: 26.96ms
prediction  13, time: 27.07ms
prediction  14, time: 27.13ms
prediction  15, time: 27.10ms

pr branch:
model_print_timings: prompt eval time =  1184.35 ms /  2023 tokens (    0.59 ms per token)
model_print_timings:        eval time =   389.48 ms /    15 runs   (   25.97 ms per token)
model_print_timings:       total time =  2577.57 ms
========== eval time log of each prediction ==========
prediction   0, time: 1184.35ms
prediction   1, time: 27.49ms
prediction   2, time: 26.47ms
prediction   3, time: 25.85ms
prediction   4, time: 25.77ms
prediction   5, time: 26.20ms
prediction   6, time: 25.94ms
prediction   7, time: 25.72ms
prediction   8, time: 25.70ms
prediction   9, time: 26.19ms
prediction  10, time: 25.67ms
prediction  11, time: 25.65ms
prediction  12, time: 25.61ms
prediction  13, time: 25.93ms
prediction  14, time: 25.69ms
prediction  15, time: 25.61ms

File: intel_extension_for_transformers/llm/library/jblas/.clang-format [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Language:        Cpp
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/library/jblas/.clang-format:1]: is this file necessary?

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 2]
```diff
@@ -511,54 +520,71 @@ int Model::quant_model(const std::string& model_path, const std::string& out_pat
 }
 
 size_t Model::jblas_qpack(const int8_t* src_w, const float* src_scales, const int8_t* src_zps, void* dstpr,
-                          const quant_params_internal params, int nthread, int n, int k) {
-  using CompType = jblas::prologue::weight_comp::gemm_kblcok::PrologueBIDs;
-  using namespace ne_jblas;
-  auto cd = jblas::utils::parallel::CpuDevice::getInstance();
+                          const quant_params_internal params, int nthread, int N, int K) {
+  
+  auto constexpr ISA = jblas::gemm::ICoreRowNAvx512vnniKBlock<48, 4>::ISA;
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:525]: bound to single ISA?

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 3]
```diff
@@ -511,54 +520,71 @@ int Model::quant_model(const std::string& model_path, const std::string& out_pat
...
+  using Launcher = jblas::wrapper::gemm::LauncherIntKBlock<ISA, jblas::gemm::ICoreRowNAvx512vnniKBlock<48, 4>,
+                                              jblas::prologue_a::gemm::ActivationF32KBlockQuantize,
+                                              jblas::prologue_b::gemm::WeightKBlockNInteger,
+                                              jblas::epilogue::gemm::AccumulatorWriteBackFp32>;
+  using Parallel = jblas::parallel::gemm::SchedulerKBlockS<jblas::gemm::ICoreRowNAvx512vnniKBlock<48, 4>>;
+  Launcher launcher;
+  int blocksize = params.group_size;
+  bool isAsym = params.alg == quant_alg::asym;
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:533]:
```suggestion
  bool isAsym = src_zps == nullptr;
```

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 4]
```diff
@@ -229,71 +239,60 @@ using ScaleWriteBackS32S8 = ScaleWriteBack<ISA_T, int32_t, int8_t>;
...
     deserializeBuffer(buf, true);
-    jblas::prologue::ISerialBuffer::deserializeBuffer(buf, true);
-    mWPtr = get<T>();
-    mWSize = size<T>();
+    mWBuf.deserializeBuffer(buf, true);
   }
 
-  virtual void serialize(int8_t* wptr) {
-    jblas::prologue::gemm::WeightBase::serializeToBuffer(wptr);
+  void serialize(int8_t* wptr) override {
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:271]: need activation a serialization method? when will we serialize these?

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 5]
```diff
@@ -303,109 +302,76 @@ class PackedWeightBatch : public jblas::prologue::gemm::WeightBase, public jblas
...
+    const int ldb;
+    const StorageType* packedW;
   };
 
   JBLAS_CODE getWeight(...) = delete;
 
   JBLAS_CODE getWeight(WType** dstptr, int* dststep, int /* b_size */, int /* k_size */, int /* n_size */, int b_offset,
-                       int k_offset, int n_offset, const jblas::prologue::gemm::WeightBase* ptr) {
-    const auto wptr = dynamic_cast<const PW_T*>(ptr);
+                       int k_offset, int n_offset, const Param& param, void* /* tmpcache */, size_t /* cachesize */) {
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:318]: b_size not used?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp [hunk 6]
```diff
@@ -868,137 +870,50 @@ quant_params_internal quant_params_to_internal(const quant_params& params) {
...
-      static KernelRef kernelref;
-      auto packedw = kernel.createStorage(n, k, params.group_size, params.alg == quant_alg::asym);
-      packedw.assign(dstbptr);
-      if (cd->AMX_BF16()) {
-        kernel.packTransposeWeight(n, k, f32ptr, k, &packedw);
-      } else {
-        kernelref.packTransposeWeight(n, k, f32ptr, k, &packedw);
-      }
-      return packedw.mSize;
+      printf("Compute Int8 is not supported by float quant types, will be igonred\n");
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp:896]: I think current jblas has supported

**Comment by airMeng** [comment]
shall there be any updated on jblas doc https://github.com/intel/intel-extension-for-transformers/blob/refactor_itrex_backend_based_on_new_jblas/intel_extension_for_transformers/llm/library/jblas/README.md ?

@luoyu-intel @zhewang1-intc

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 5 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:318]: Yes. All sizes are not used as this WeightPack class was designed to packed with paddings and should never be runtime copy and padded. Therefore, a pointer to the storage (with some offset) is given directly and the `b/k/n_size` and the `tmpcache` parameters are just here for a consistent API.

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 4 above)
Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:271]: assert false looks better for me

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 2 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:599]: this function is used by gptq, I solved it in gptq branch.

File: intel_extension_for_transformers/llm/operator/csrc/dispatcher/include/jblas_customop.hpp [hunk 7]
```diff
@@ -50,7 +48,7 @@ class AlphaBetaProcess {
     float alpha, beta;
   };
   JBLAS_CODE forward(float* cacheptr, const int cachestep, const int M_offset, const int N_offset, const int M,
-                     const int N, const Param& _param) {
+                     const int N, const Param& _param, void* tmpcache = nullptr, size_t cachesize = -1) {
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/operator/csrc/dispatcher/include/jblas_customop.hpp:51]: seems not use tmpcache  and cachesize ?

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c [hunk 8]
```diff
@@ -130,7 +130,7 @@ static_assert(sizeof(block_q8_1) == 2 * sizeof(float) + QK8_1, "wrong q8_1 block
 #if UINTPTR_MAX == 0xFFFFFFFF
 #define NE_MEM_ALIGN 4
 #else
-#define NE_MEM_ALIGN 16
+#define NE_MEM_ALIGN 64
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:133]: why we change mem align from 16 to 64 ?

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/jblas_gemm.h [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+//  limitations under the License.
+
+/*++
+Module Name:
+
+    jblas_gemm.h
+
+Abstract:
+
+    C APIs of BesTLA GEMMs.
```
[the remark this answers is not in the document]
Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/jblas_gemm.h:22]: yes, maybe in neural speed

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c (hunk 8 above)
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:133]: 16-byte is not useful to Intel CPUs. Always padded to 64 bytes, and no need to consider 32-bit system which our kernel does not support.

File: intel_extension_for_transformers/llm/library/jblas/.clang-format (hunk 1 above)
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/library/jblas/.clang-format:1]: multiple .clang-format files are not a problem.

**Comment by VincyZhang** [comment]
https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/200/

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 2 above)
Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:599]: this pr https://github.com/intel/intel-extension-for-transformers/pull/791 ?

Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:599]: not this, I will push new PR after this one has been merged.

**Comment by VincyZhang** [comment]
https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/201/

**Comment by VincyZhang** [comment]
> https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/200/

Some model failed without output, just hangs for days (not due to disk), please check @zhewang1-intc @luoyu-intel

**Comment by VincyZhang** [comment]
https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/203/

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 4 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:276]: Will assert false in a following PR so that the implementation for the new jblas will in git history and (be restored easily if needed).

**Review by luoyu-intel — APPROVED** [review]

**Review by airMeng — APPROVED** [review]

**Comment by VincyZhang** [comment]
https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/204/

**Comment by VincyZhang** [comment]
Neuralchat UT failure is irrelevant, please check @lvliang-intel
